lib/bankscrap/ing/bank.rb

```diff
@@ -140,64 +139,58 @@ def post_auth(ticket)
         post(POST_AUTH_ENDPOINT, fields: params)
       end
 
-      def save_pinpad_numbers(pinpad)
-        current_pinpad_paths = []
-        pinpad.each_with_index do |digit, index|
-          tmp = Tempfile.new(["pinpad_number_#{index}", '.png'])
-          File.open(tmp.path, 'wb') { |f| f.write(Base64.decode64(digit)) }
-          current_pinpad_paths << tmp.path
+      def extract_images_from_pinpad(pinpad)
+        pinpad.map do |string|
+          extract_info_from_encoded_png(string)
         end
-
-        current_pinpad_paths
       end
 
-      def recognize_pinpad_numbers(current_pinpad_paths)
-        real_numbers = []
-        current_numbers = Magick::ImageList.new(*current_pinpad_paths)
+      def extract_info_from_encoded_png(encoded_png)
+        decoded = Base64.decode64(encoded_png)
+        # PNG format first bytes are headers that we can discard. We know where the data chunk starts,
+        # so we can directly read the image data. See: https://www.w3.org/TR/PNG-Structure.html
+        length  = decoded.slice(33, 4).unpack('L>').first
+        data    = decoded.slice(41, length)
+        Zlib::Inflate.inflate(data)       # We decompress the data
+                     .delete("\u0000")    # Levenshtein distance doesn't work with null bytes
+                     .slice(2_000, 5_000) # The whole image is too slow to compare, we can use a chunk in the middle.
```

Comment:
This is so cool. I was wondering, why from 2,000 to 5,000? Does it work for every number always no matter what?